mesonbuild/coredata.py

```diff
@@ -160,10 +160,16 @@ def __init__(self, description, choices: T.List[str], value, yielding=None):
 
     def validate_value(self, value):
         if value not in self.choices:
+            if isinstance(value, bool):
+                _type = 'boolean'
+            elif isinstance(value, (int, float)):
+                _type = 'number'
+            else:
+                _type = 'string'
             optionsstring = ', '.join(['"%s"' % (item,) for item in self.choices])
-            raise MesonException('Value "{}" for combo option "{}" is not one of the choices.'
-                                 ' Possible choices are: {}.'.format(
-                                     value, self.description, optionsstring))
+            raise MesonException('Value "{}" (of type "{}") for combo option "{}" is not one of the choices.'
+                                 ' Possible choices are (as string): {}.'.format(
+                                     value, _type, self.description, optionsstring))
```

Comment:
Any type other than str is forbidden atm. I would just move the check from `__init__()` into `validate_value()`:

```
            if not isinstance(value, str):
                raise MesonException('Combo choice elements must be strings.')
```